File: cpp/rotate_array.cc
/*
 * Filename: rotate_array.cc
 * Author: Jun Xu
 * Email: jxu.vimacs@gmail.com
 * Created Time: Fri 14 Aug 2015 12:24:04 PM EDT
 */

#include <iostream>
using namespace std;

class Solution {
	public:
		void rotate(vector<int>& nums, int k) {
			if (nums.empty())
			{
				return;
			}

			int n = nums.size();
			k %= n;

			if (!k)
			{
				return;
			}

			reverse(nums, 0, n-k-1);
			reverse(nums, n-k, n-1);
			reverse(nums, 0, n-1);
		}

	private:
		void reverse(
				vector<int>& nums,
				int i,
				int j)
		{
			while (i < j)
			{
				int tmp = nums[i];
				nums[i] = nums[j];
				nums[j] = tmp;
				i++;
				j--;
			}
		}
};

File: java/RotateArray.java
/*
 * Filename: RotateArray.java
 * Author: Jun Xu
 * Email: jxu.vimacs@gmail.com
 * Created Time: Tue 07 Jul 2015 04:44:35 PM EDT
 */

public class Solution {
	public void rotate(int[] nums, int k) {
		if (nums == null || nums.length == 0) {
			return;
		}

		k = k % nums.length;
		if (k == 0) {
			return;
		}

		reverse(nums, 0, nums.length - 1 );
		reverse(nums, 0, k - 1 );
		reverse(nums, k, nums.length - 1 );
	}

	private void reverse(int[] nums, int stt, int end) {
		while (stt < end) {
			int temp = nums[stt];
			nums[stt] = nums[end];
			nums[end] = temp;
			stt++;
			end--;
		}
	}
}
